sockeye/decoder.py

```diff
@@ -1259,3 +1286,160 @@ def state_shapes(self,
     def get_max_seq_len(self) -> Optional[int]:
         #  The positional embeddings potentially pose a limit on the maximum length at inference time.
         return self.pos_embedding.get_max_seq_len()
+
+
+class CustomSeqDecoderConfig(Config):
+    def __init__(self, decoder_layers: List[layers.LayerConfig], num_embed: int, dtype: str = C.DTYPE_FP32):
+        super().__init__()
+        self.num_embed = num_embed
+        self.decoder_layers = decoder_layers
+        self.dtype = dtype
+
+
+@Decoder.register(CustomSeqDecoderConfig, C.CUSTOM_SEQ_TYPE)
+class CustomSeqDecoder(Decoder):
+    """
+    Decoder composed of a customizable list of building blocks/layers, such as RNNs, CNNs and attention mechanisms.
+    """
+
+    def __init__(self, config: CustomSeqDecoderConfig,
+                 prefix: str = C.DECODER_PREFIX) -> None:
+        super().__init__(config.dtype)
+        self.config = config
+        self.prefix = prefix
+        self.layers = []
+        input_num_hidden = config.num_embed
+        for idx, layer_config in enumerate(config.decoder_layers):
```

Comment:
in custom_seq_parser we make sure that at least one source attention layer is present, so technically it is indeed not possible. If it was empty `self.layers` would be empty. Everything would work except that there is no connection between the encoder and the decoder side, which would lead to an MXNet issue.